In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
tab1 = pd.read_excel('C:/Users/mrinm/OneDrive/Desktop/Vaccination-Project---Python-SQl-Power-BI/coverage-data.xlsx')


In [3]:
tab2 = pd.read_excel('incidence-rate-data.xlsx')
tab3 = pd.read_excel('reported-cases-data.xlsx')
tab4 = pd.read_excel('vaccine-introduction-data.xlsx')
tab5 = pd.read_excel('vaccine-schedule-data.xlsx')


In [4]:
tab1.head()

GROUP CODE   NAME    YEAR  ANTIGEN  \
0  COUNTRIES  ABW  Aruba  2023.0      BCG   
1  COUNTRIES  ABW  Aruba  2023.0      BCG   
2  COUNTRIES  ABW  Aruba  2023.0  DIPHCV4   
3  COUNTRIES  ABW  Aruba  2023.0  DIPHCV4   
4  COUNTRIES  ABW  Aruba  2023.0  DIPHCV5   

                                 ANTIGEN_DESCRIPTION COVERAGE_CATEGORY  \
0                                                BCG             ADMIN   
1                                                BCG          OFFICIAL   
2  Diphtheria-containing vaccine, 4th dose (1st b...             ADMIN   
3  Diphtheria-containing vaccine, 4th dose (1st b...          OFFICIAL   
4  Diphtheria-containing vaccine, 5th dose (2nd b...             ADMIN   

  COVERAGE_CATEGORY_DESCRIPTION  TARGET_NUMBER   DOSES  COVERAGE  
0       Administrative coverage            NaN     NaN       NaN  
1             Official coverage            NaN     NaN       NaN  
2       Administrative coverage         1044.0   945.0     90.52  
3             Official coverage            NaN     NaN     90.52  
4       Administrative coverage         1219.0  1008.0     82.69

In [5]:
tab2.head()

GROUP CODE   NAME    YEAR          DISEASE  \
0  COUNTRIES  ABW  Aruba  2023.0              CRS   
1  COUNTRIES  ABW  Aruba  2023.0       DIPHTHERIA   
2  COUNTRIES  ABW  Aruba  2023.0  INVASIVE_MENING   
3  COUNTRIES  ABW  Aruba  2023.0          MEASLES   
4  COUNTRIES  ABW  Aruba  2023.0            MUMPS   

              DISEASE_DESCRIPTION                     DENOMINATOR  \
0     Congenital rubella syndrome          per 10,000 live births   
1                      Diphtheria  per 1,000,000 total population   
2  Invasive meningococcal disease  per 1,000,000 total population   
3                         Measles  per 1,000,000 total population   
4                           Mumps  per 1,000,000 total population   

   INCIDENCE_RATE  
0             0.0  
1             0.0  
2             9.3  
3             NaN  
4             0.0

In [6]:
tab3.head()

GROUP CODE   NAME    YEAR          DISEASE  \
0  COUNTRIES  ABW  Aruba  2023.0              CRS   
1  COUNTRIES  ABW  Aruba  2023.0       DIPHTHERIA   
2  COUNTRIES  ABW  Aruba  2023.0  INVASIVE_MENING   
3  COUNTRIES  ABW  Aruba  2023.0          MEASLES   
4  COUNTRIES  ABW  Aruba  2023.0            MUMPS   

              DISEASE_DESCRIPTION  CASES  
0     Congenital rubella syndrome    0.0  
1                      Diphtheria    0.0  
2  Invasive meningococcal disease    1.0  
3                         Measles    NaN  
4                           Mumps    0.0

In [7]:
tab4.head()

ISO_3_CODE  COUNTRYNAME WHO_REGION    YEAR  \
0        AFG  Afghanistan       EMRO  2023.0   
1        AFG  Afghanistan       EMRO  2023.0   
2        AFG  Afghanistan       EMRO  2023.0   
3        AFG  Afghanistan       EMRO  2023.0   
4        AFG  Afghanistan       EMRO  2023.0   

                                   DESCRIPTION INTRO  
0             aP (acellular pertussis) vaccine    No  
1                          Hepatitis A vaccine    No  
2                          Hepatitis B vaccine   Yes  
3                              HepB birth dose   Yes  
4  Hib (Haemophilus influenzae type B) vaccine   Yes

In [8]:
tab5.head()

ISO_3_CODE COUNTRYNAME WHO_REGION    YEAR VACCINECODE  \
0        ABW       Aruba       AMRO  2023.0  DTAPHIBIPV   
1        ABW       Aruba       AMRO  2023.0  DTAPHIBIPV   
2        ABW       Aruba       AMRO  2023.0  DTAPHIBIPV   
3        ABW       Aruba       AMRO  2023.0  DTAPHIBIPV   
4        ABW       Aruba       AMRO  2023.0     DTAPIPV   

                VACCINE_DESCRIPTION  SCHEDULEROUNDS  TARGETPOP  \
0  DTaP-Hib-IPV (acellular) vaccine             1.0        NaN   
1  DTaP-Hib-IPV (acellular) vaccine             2.0        NaN   
2  DTaP-Hib-IPV (acellular) vaccine             3.0        NaN   
3  DTaP-Hib-IPV (acellular) vaccine             4.0    B_2YL_W   
4      DTaP-IPV (acellular) vaccine             5.0  B_CHILD_W   

  TARGETPOP_DESCRIPTION   GEOAREA AGEADMINISTERED SOURCECOMMENT  
0       General/routine  NATIONAL              M2           NaN  
1       General/routine  NATIONAL              M4           NaN  
2       General/routine  NATIONAL              M6           NaN  
3       General/routine  NATIONAL             M15           NaN  
4       General/routine  NATIONAL              Y4           NaN

#### Checking for missing Values


##### For table 1


In [9]:
# # First, drop rows where 'TARGET_NUMBER' is missing
# tab1.dropna(subset=['TARGET_NUMBER'], inplace=True)

In [10]:
tab1['ANTIGEN'].isnull().sum()

np.int64(1)

In [11]:
tab1['ANTIGEN_DESCRIPTION'].isnull().sum()

np.int64(1)

In [12]:
tab1['CODE'].isnull().sum()

np.int64(1)

In [13]:
tab1['COVERAGE'].isnull().sum()

np.int64(169382)

In [14]:
tab1['COVERAGE_CATEGORY'].isnull().sum()

np.int64(1)

In [15]:
tab1['COVERAGE_CATEGORY_DESCRIPTION'].isnull().sum()

np.int64(1)

In [16]:
tab1['DOSES'].isnull().sum()

np.int64(320532)

In [17]:
tab1['GROUP'].isnull().sum()

np.int64(0)

In [18]:
tab1['NAME'].isnull().sum()

np.int64(1275)

In [19]:
tab1['YEAR'].isnull().sum()

np.int64(1)

In [20]:
tab1.head(5)

GROUP CODE   NAME    YEAR  ANTIGEN  \
0  COUNTRIES  ABW  Aruba  2023.0      BCG   
1  COUNTRIES  ABW  Aruba  2023.0      BCG   
2  COUNTRIES  ABW  Aruba  2023.0  DIPHCV4   
3  COUNTRIES  ABW  Aruba  2023.0  DIPHCV4   
4  COUNTRIES  ABW  Aruba  2023.0  DIPHCV5   

                                 ANTIGEN_DESCRIPTION COVERAGE_CATEGORY  \
0                                                BCG             ADMIN   
1                                                BCG          OFFICIAL   
2  Diphtheria-containing vaccine, 4th dose (1st b...             ADMIN   
3  Diphtheria-containing vaccine, 4th dose (1st b...          OFFICIAL   
4  Diphtheria-containing vaccine, 5th dose (2nd b...             ADMIN   

  COVERAGE_CATEGORY_DESCRIPTION  TARGET_NUMBER   DOSES  COVERAGE  
0       Administrative coverage            NaN     NaN       NaN  
1             Official coverage            NaN     NaN       NaN  
2       Administrative coverage         1044.0   945.0     90.52  
3             Official coverage            NaN     NaN     90.52  
4       Administrative coverage         1219.0  1008.0     82.69

Also we are having 1275 missing values for the column 'NAME'. 


The most reliable method is to create a separate, clean reference table of Code and Name from your dataset.

In [21]:
# Create a mapping of Code to Name
code_to_name_mapping = tab1[['CODE','NAME']].dropna().drop_duplicates().set_index('CODE')['NAME']

# Fill missing Name values using the mapping
tab1['NAME'] = tab1['CODE'].map(code_to_name_mapping).fillna(tab1['NAME'])

Cannot find the names with code. So instead just setting the null 'NAME' values as 'Unknown Country'

In [22]:
tab1['NAME'] = tab1['NAME'].fillna('Unknown Country')

In [23]:
# # Impute missing 'COVERAGE' values with the median for the same Antigen and Year

# grouped_data = tab1.groupby(['ANTIGEN', 'YEAR'])
# coverage_groups = grouped_data['COVERAGE']
# group_medians = coverage_groups.median()


# def fill_with_median(x):
#     return x.fillna(x.median())

# # Now applying this function to each group
# imputed_coverage = coverage_groups.transform(fill_with_median)

# # Finally, assiging the new, imputed values back to the original column
# tab1['COVERAGE'] = imputed_coverage

In [24]:
tab1['COVERAGE'].isna().sum()

np.int64(169382)

In [25]:
# # Impute missing 'DOSES' values with the median for the same Antigen and Year

# grouped_data1 = tab1.groupby(['ANTIGEN', 'YEAR'])
# coverage_from_grouped_data1 = grouped_data1['DOSES']
# group_medians1 = coverage_from_grouped_data1.median()

# # Now applying this function to each group
# imputed_doses = coverage_from_grouped_data1.transform(fill_with_median)

# # Finally, assiging the new, imputed values back to the original column
# tab1['DOSES'] = imputed_doses


In [26]:
tab1['DOSES'].isna().sum()

np.int64(320532)

In [27]:
## 1. Clean and standardize column names
tab1.rename(columns={
    'Group': 'GROUP', 'Code': 'CODE', 'Name': 'NAME', 'Year': 'YEAR', 'Antigen': 'ANTIGEN',
    'Antigen_description': 'ANTIGEN_DESCRIPTION', 'Coverage_category': 'COVERAGE_CATEGORY',
    'Coverage_category_description': 'COVERAGE_CATEGORY_DESCRIPTION',
    'Target number': 'TARGET_NUMBER', 'Dodge': 'DOSES', 'Coverage': 'COVERAGE'
}, inplace=True)

## 2. Handle missing values and data type conversions
# Drop rows where a key column like 'TARGET_NUMBER' is missing
tab1.dropna(subset=['TARGET_NUMBER'], inplace=True)

# Impute missing values in COVERAGE and DOSES using group-based medians
tab1['COVERAGE'] = tab1.groupby(['ANTIGEN', 'YEAR'])['COVERAGE'].transform(lambda x: x.fillna(x.median()))
tab1['DOSES'] = tab1.groupby(['ANTIGEN', 'YEAR'])['DOSES'].transform(lambda x: x.fillna(x.median()))

# Drop any remaining rows with missing COVERAGE or DOSES after imputation
tab1.dropna(subset=['COVERAGE', 'DOSES'], inplace=True)

# Handle data type conversions
tab1['TARGET_NUMBER'] = tab1['TARGET_NUMBER'].astype(int).astype('Int64')
tab1['DOSES'] = tab1['DOSES'].astype(int).astype('Int64')
tab1['YEAR'] = tab1['YEAR'].astype(int)

# tab1['NAME'] = tab1['NAME'].fillna('Unknown Country')

# Drop the rows which are having missing values 
tab1 = tab1.dropna()

##### For table 2


In [28]:
tab2['CODE'].isnull().sum()

np.int64(1)

In [29]:
tab2['DENOMINATOR'].isnull().sum()

np.int64(1)

In [30]:
tab2['DISEASE'].isnull().sum()

np.int64(1)

In [31]:
tab2['DISEASE_DESCRIPTION'].isnull().sum()

np.int64(1)

In [32]:
tab2['GROUP'].isnull().sum()

np.int64(0)

In [33]:
tab2['INCIDENCE_RATE'].isnull().sum()

np.int64(23362)

In [34]:
tab2['NAME'].isnull().sum()

np.int64(1)

In [35]:
tab2['YEAR'].isnull().sum()

np.int64(1)

In [36]:
tab2.rename(columns={
    'Group': 'GROUP',
    'Code': 'CODE',
    'Name': 'NAME',
    'Year': 'YEAR',
    'Disease': 'DISEASE',
    'Disease description': 'DISEASE_DESCRIPTION',
    'Denominator': 'DENOMINATOR',
    'Incidence rate': 'INCIDENCE_RATE'
}, inplace=True)

## 2. Handle missing values and data type conversions
# Impute missing values in INCIDENCE_RATE with the median
median_incindence_rate = tab2['INCIDENCE_RATE'].median()
tab2['INCIDENCE_RATE'].fillna(median_incindence_rate, inplace=True)


# Drop rows with missing values in critical identifier columns
tab2.dropna(subset=['CODE', 'NAME', 'YEAR'], inplace=True)
tab2 = tab2.dropna()

# Convert YEAR column to integer
tab2['YEAR'] = tab2['YEAR'].astype(int)


C:\Users\mrinm\AppData\Local\Temp\ipykernel_16276\2164821578.py:15: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  tab2['INCIDENCE_RATE'].fillna(median_incindence_rate, inplace=True)


##### For table 3

In [37]:
tab3.shape

(84870, 7)

In [38]:
tab3.head(5)

GROUP CODE   NAME    YEAR          DISEASE  \
0  COUNTRIES  ABW  Aruba  2023.0              CRS   
1  COUNTRIES  ABW  Aruba  2023.0       DIPHTHERIA   
2  COUNTRIES  ABW  Aruba  2023.0  INVASIVE_MENING   
3  COUNTRIES  ABW  Aruba  2023.0          MEASLES   
4  COUNTRIES  ABW  Aruba  2023.0            MUMPS   

              DISEASE_DESCRIPTION  CASES  
0     Congenital rubella syndrome    0.0  
1                      Diphtheria    0.0  
2  Invasive meningococcal disease    1.0  
3                         Measles    NaN  
4                           Mumps    0.0

In [39]:
tab3['CASES'].isnull().sum()

np.int64(19400)

We cannot replace with median or mean value for the cases column because the median would be 0 if we calculate and so it will mean that 19,400 countries are having 0 reported cases which is not correct at all and can lead to bias. 

So its better to drop the rows with missing values


In [40]:
# tab3 = tab3.dropna(subsets=['CASES'],inplace=True)

In [41]:
tab3['CODE'].isnull().sum()

np.int64(1)

In [42]:
tab3['DISEASE'].isnull().sum()

np.int64(1)

In [43]:
tab3['DISEASE_DESCRIPTION'].isnull().sum()

np.int64(1)

In [44]:
tab3['GROUP'].isnull().sum()

np.int64(0)

In [45]:
tab3['NAME'].isnull().sum()

np.int64(1)

In [46]:
tab3['YEAR'].isnull().sum()

np.int64(1)

In [ ]:
## 1. Clean and standardize column names
tab3.rename(columns={
    'Group': 'GROUP',
    'Code': 'CODE',
    'Name': 'NAME',
    'Year': 'YEAR',
    'Disease': 'DISEASE',
    'Disease description': 'DISEASE_DESCRIPTION',
    'Cases': 'CASES'
}, inplace=True)

# Drop rows where the columns has a missing value
tab3.dropna(subset=['CASES','YEAR','NAME','DISEASE_DESCRIPTION','DISEASE','CODE'],inplace=True)

# Handle data type conversion
tab3['YEAR'] = tab3['YEAR'].astype(int)

##### For Table 4

In [48]:
tab4.head(5)

ISO_3_CODE  COUNTRYNAME WHO_REGION    YEAR  \
0        AFG  Afghanistan       EMRO  2023.0   
1        AFG  Afghanistan       EMRO  2023.0   
2        AFG  Afghanistan       EMRO  2023.0   
3        AFG  Afghanistan       EMRO  2023.0   
4        AFG  Afghanistan       EMRO  2023.0   

                                   DESCRIPTION INTRO  
0             aP (acellular pertussis) vaccine    No  
1                          Hepatitis A vaccine    No  
2                          Hepatitis B vaccine   Yes  
3                              HepB birth dose   Yes  
4  Hib (Haemophilus influenzae type B) vaccine   Yes

In [49]:
tab4['COUNTRYNAME'].isna().sum()

np.int64(1)

In [50]:
tab4['DESCRIPTION'].isna().sum()

np.int64(1)

In [51]:
tab4['INTRO'].isna().sum()

np.int64(1)

In [52]:
tab4['ISO_3_CODE'].isna().sum()

np.int64(0)

In [53]:
tab4['WHO_REGION'].isna().sum()

np.int64(1)

In [54]:
tab4['YEAR'].isna().sum()

np.int64(1)

In [55]:
## 1. Clean and standardize column names
tab4.rename(columns={
    'ISO_3_CODE': 'CODE',
    'COUNTRYNAME': 'NAME',
    'WHO_REGION': 'WHO_REGION',
    'YEAR': 'YEAR',
    'DESCRIPTION': 'DESCRIPTION',
    'INTRO': 'INTRO'
}, inplace=True)

## 2. Handling missing values and data type conversions
tab4 = tab4.dropna()

# Converting YEAR column to integer
tab4['YEAR'] = tab4['YEAR'].astype(int)

##### For Table 5

In [56]:
tab5.shape

(8053, 12)

In [57]:
tab5['AGEADMINISTERED'].isna().sum()

np.int64(1046)

In [58]:
tab5['AGEADMINISTERED'].value_counts()

AGEADMINISTERED
M2             434
M4             434
1st contact    362
M6             345
+M6            341
              ... 
Y8               1
Y49              1
Y1-Y45           1
Y7-Y11           1
B-M24            1
Name: count, Length: 425, dtype: int64

In [59]:
tab5['COUNTRYNAME'].isna().sum()

np.int64(1)

In [60]:
tab5['GEOAREA'].isna().sum()

np.int64(31)

In [61]:
tab5['ISO_3_CODE'].isna().sum()

np.int64(0)

In [62]:
tab5['SCHEDULEROUNDS'].isna().sum()

np.int64(1)

In [63]:
tab5['SOURCECOMMENT'].isna().sum()

np.int64(2914)

In [64]:
tab5['SOURCECOMMENT'].value_counts()

SOURCECOMMENT
0-11 mois                                                                                                               185
Pregnant women                                                                                                          157
Health and care workers                                                                                                 124
Children with chronic conditions                                                                                         97
0-11 months                                                                                                              91
                                                                                                                       ... 
outrbreak response                                                                                                        1
Girls aged 9-14 years old (in school and out of school) reached through an HPV Multi Age Cohort Vaccination Campaign  

In [65]:
tab5['TARGETPOP'].isna().sum()

np.int64(4258)

In [66]:
tab5['TARGETPOP'].value_counts()

TARGETPOP
RISKGROUPS    1247
ADULTS         658
PW             535
HW             330
TRAVELLERS     208
B_CHILD_W      164
FEMALE         152
B_ADO_W        130
B_2YL_W        125
BOTH           124
CATCHUP_C       84
CATCHUP_A       19
B_ADO_F         10
B_2YL_F          6
B_CHILD_F        3
Name: count, dtype: int64

In [67]:
tab5['TARGETPOP_DESCRIPTION'].isna().sum()

np.int64(1)

In [68]:
tab5['VACCINECODE'].isna().sum()

np.int64(1)

In [69]:
tab5['VACCINE_DESCRIPTION'].isna().sum()

np.int64(1)

In [70]:
tab5['WHO_REGION'].isna().sum()

np.int64(1)

In [71]:
tab5['YEAR'].isna().sum()

np.int64(1)

In [72]:
## 1. Clean and standardize column names
tab5.rename(columns={
    'ISO_3_CODE': 'CODE',
    'COUNTRYNAME': 'NAME',
    'Who Region': 'WHO_REGION',
    'Year': 'YEAR',
    'Vaccine code': 'VACCINECODE',
    'Vaccine description': 'VACCINE_DESCRIPTION',
    'Schedule rounds': 'SCHEDULEROUNDS',
    'Target pop': 'TARGETPOP',
    'Target pop description': 'TARGETPOP_DESCRIPTION',
    'Geoarea': 'GEOAREA',
    'Age administered': 'AGEADMINISTERED',
    'Source comment': 'SOURCECOMMENT'
}, inplace=True)

## 2. Handling missing values and data type conversions
# Droping the SOURCECOMMENT column as it has many missing values and is not critical
# tab5 = tab5.drop(columns=['SOURCECOMMENT'], axis=1)

# Imputing missing values for AGEADMINISTERED and TARGETPOP with 'Unknown'
tab5['AGEADMINISTERED'] = tab5['AGEADMINISTERED'].fillna('Unknown')
tab5['TARGETPOP'] = tab5['TARGETPOP'].fillna('Unknown')

# Droping rows where core identifier columns have missing values
tab5 = tab5.dropna()

# Converting YEAR column to integer
tab5['YEAR'] = tab5['YEAR'].astype(int)

Now We need solve the problem of NAME column in table1 

In [73]:
# Concatenate all CODE and NAME pairs into one DataFrame
all_tabs = [tab1, tab2, tab3, tab4, tab5]

code_name_map_df = pd.DataFrame()
for tab in all_tabs:
    # Check if the required columns exist before concatenating
    if 'CODE' in tab.columns and 'NAME' in tab.columns:
        temp_df = tab[['CODE', 'NAME']].dropna().drop_duplicates()
        code_name_map_df = pd.concat([code_name_map_df, temp_df])

# Create a clean, final mapping with no duplicates
code_name_map = code_name_map_df.drop_duplicates(subset=['CODE']).set_index('CODE')['NAME']

# Use the comprehensive mapping to fill missing 'NAME' values in tab1
tab1['NAME'] = tab1['CODE'].map(code_name_map).fillna(tab1['NAME'])

In [74]:
tab1['NAME'].isna().sum()

np.int64(0)

As you can see in the above code cell that there are no empty rows in the NAME Column

In [75]:
import pyodbc
from sqlalchemy import create_engine

def create_sql_engine():
    try:
        # Define the parts of your connection string
        server = 'Mrin_2130115\\MySQL2002'
        database = 'VACCINATION_PROJECT'
        driver = 'ODBC Driver 17 for SQL Server'
        
        # Build the connection string using SQLAlchemy's standard URL format
        conn_str = f"mssql+pyodbc:///?odbc_connect=DRIVER%3D%7B{driver}%7D%3BSERVER%3D{server}%3BDATABASE%3D{database}%3BTRUSTED_CONNECTION%3Dyes%3B"
        
        # Create the engine
        engine = create_engine(conn_str)
        print("SQLAlchemy engine created successfully.")
        return engine
    except Exception as ex:
        print(f"Error creating SQLAlchemy engine: {ex}")
        return None
    
# Main execution block remains the same
if __name__ == '__main__':
    engine = create_sql_engine()
    if engine:
        # Pass the engine to your create_tables() function
        # create_tables(engine)
        pass

SQLAlchemy engine created successfully.


So Now we will create the tables and to execute the commands in SQL we will use cursor function

-- Create a Countries table to avoid redundancy
CREATE TABLE Countries (
    Code VARCHAR(3) PRIMARY KEY,
    Name VARCHAR(100),
    WhoRegion VARCHAR(50)
);

-- Table 1: CoverageData (references Countries table)
CREATE TABLE CoverageData (
    CoverageID INT PRIMARY KEY IDENTITY(1,1),
    Code VARCHAR(3) FOREIGN KEY REFERENCES Countries(Code),
    Year INT,
    Antigen VARCHAR(20),
    Antigen_description VARCHAR(100),
    Coverage_category VARCHAR(50),
    Coverage_category_description VARCHAR(100),
    Target_number BIGINT,
    Doses BIGINT,
    Coverage FLOAT
);

-- Table 2: IncidenceRateData (references Countries table)
CREATE TABLE IncidenceRateData (
    IncidenceID INT PRIMARY KEY IDENTITY(1,1),
    Code VARCHAR(3) FOREIGN KEY REFERENCES Countries(Code),
    Year INT,
    Disease VARCHAR(50),
    Disease_description VARCHAR(100),
    Incidence_rate FLOAT
);

-- Table 3: ReportedCasesData (references Countries table)
CREATE TABLE ReportedCasesData (
    CasesID INT PRIMARY KEY IDENTITY(1,1),
    Code VARCHAR(3) FOREIGN KEY REFERENCES Countries(Code),
    Year INT,
    Disease VARCHAR(50),
    Disease_description VARCHAR(100),
    Cases FLOAT
);

-- Table 4: VaccineIntroduction (references Countries table)
CREATE TABLE VaccineIntroduction (
    IntroID INT PRIMARY KEY IDENTITY(1,1),
    Code VARCHAR(3) FOREIGN KEY REFERENCES Countries(Code),
    Year INT,
    Description VARCHAR(100),
    Intro VARCHAR(10)
);

-- Table 5: VaccineScheduleData (references Countries table)
CREATE TABLE VaccineScheduleData (
    ScheduleID INT PRIMARY KEY IDENTITY(1,1),
    Code VARCHAR(3) FOREIGN KEY REFERENCES Countries(Code),
    Year INT,
    VaccineCode VARCHAR(50),
    VaccineDescription VARCHAR(100),
    ScheduleRounds FLOAT,
    TARGETPOP VARCHAR(50),
    TARGETPOP_DESCRIPTION VARCHAR(100),
    GEOAREA VARCHAR(50),
    AGEADMINISTERED VARCHAR(40)
);

In [76]:
def create_tables(engine):
    """Creates the specified tables in the SQL Server database."""
    # Get a raw pyodbc connection from the SQLAlchemy engine
    conn = engine.raw_connection()
    try:
        with conn.cursor() as cursor:
            # Create a Countries table to avoid redundancy
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS Countries (
                    Code VARCHAR(3) PRIMARY KEY,
                    Name VARCHAR(100),
                    WhoRegion VARCHAR(50)
                );
            """)
            print("Table 'Countries' created.")

            # Table 1: CoverageData (references Countries table)
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS CoverageData (
                    CoverageID INT PRIMARY KEY IDENTITY(1,1),
                    Code VARCHAR(3) FOREIGN KEY REFERENCES Countries(Code),
                    Year INT,
                    Antigen VARCHAR(20),
                    Antigen_description VARCHAR(100),
                    Coverage_category VARCHAR(50),
                    Coverage_category_description VARCHAR(100),
                    Target_number BIGINT,
                    Doses BIGINT,
                    Coverage FLOAT
                );
            """)
            print("Table 'CoverageData' created.")

            # Table 2: IncidenceRateData (references Countries table)
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS IncidenceRateData (
                    IncidenceID INT PRIMARY KEY IDENTITY(1,1),
                    Code VARCHAR(3) FOREIGN KEY REFERENCES Countries(Code),
                    Year INT,
                    Disease VARCHAR(50),
                    Disease_description VARCHAR(100),
                    Incidence_rate FLOAT
                );
            """)
            print("Table 'IncidenceRateData' created.")

            # Table 3: ReportedCasesData (references Countries table)
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS ReportedCasesData (
                    CasesID INT PRIMARY KEY IDENTITY(1,1),
                    Code VARCHAR(3) FOREIGN KEY REFERENCES Countries(Code),
                    Year INT,
                    Disease VARCHAR(50),
                    Disease_description VARCHAR(100),
                    Cases FLOAT
                );
            """)
            print("Table 'ReportedCasesData' created.")

            # Table 4: VaccineIntroduction (references Countries table)
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS VaccineIntroduction (
                    IntroID INT PRIMARY KEY IDENTITY(1,1),
                    Code VARCHAR(3) FOREIGN KEY REFERENCES Countries(Code),
                    Year INT,
                    Description VARCHAR(100),
                    Intro VARCHAR(10)
                );
            """)
            print("Table 'VaccineIntroduction' created.")

            # Table 5: VaccineScheduleData (references Countries table)
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS VaccineScheduleData (
                    ScheduleID INT PRIMARY KEY IDENTITY(1,1),
                    Code VARCHAR(3) FOREIGN KEY REFERENCES Countries(Code),
                    Year INT,
                    VaccineCode VARCHAR(50),
                    VaccineDescription VARCHAR(100),
                    ScheduleRounds FLOAT,
                    TARGETPOP VARCHAR(50),
                    TARGETPOP_DESCRIPTION VARCHAR(100),
                    GEOAREA VARCHAR(50),
                    AGEADMINISTERED VARCHAR(40)
                );
            """)
            print("Table 'VaccineScheduleData' created.")

        conn.commit()  # Make sure to commit the changes to save the tables
        print("All tables created successfully.")

    except pyodbc.Error as ex:
        sqlstate = ex.args[0]
        print(f"Error creating tables: {sqlstate}")
        conn.rollback() # Rollback in case of an error
    finally:
        if conn:
            conn.close()

# The main execution block also needs to be updated to pass the engine object.
if __name__ == "__main__":
    engine = create_sql_engine()
    if engine:
        create_tables(engine) # Pass the engine object
    else:
        print("Could not establish a database connection. Tables were not created.")

SQLAlchemy engine created successfully.
Error creating tables: 42000


In [79]:
# We have to create the countries dataframe which is used as a foriegn key reference in all the tables 

all_tabs = [tab1, tab2, tab3, tab4, tab5]

countries_df = pd.DataFrame()


# Create a DataFrame of unique CODE and NAME from all tabs
unique_countries = pd.concat([
    tab1[['CODE', 'NAME']].dropna().drop_duplicates(),
    tab2[['CODE', 'NAME']].dropna().drop_duplicates(),
    tab3[['CODE', 'NAME']].dropna().drop_duplicates(),
    tab4[['CODE', 'NAME']].dropna().drop_duplicates(),
    tab5[['CODE', 'NAME']].dropna().drop_duplicates()
]).drop_duplicates(subset=['CODE'])

# Create a DataFrame of CODE and WHO_REGION from the tables that have this data
who_regions = pd.concat([
    tab4[['CODE', 'WHO_REGION']].dropna().drop_duplicates(),
    tab5[['CODE', 'WHO_REGION']].dropna().drop_duplicates()
]).drop_duplicates(subset=['CODE'])

# Merge the two DataFrames to create the final countries_df
countries_df = pd.merge(
    unique_countries,
    who_regions,
    on='CODE',
    how='left'
)

# Fill any remaining missing WHO_REGION values
countries_df['WHO_REGION'].fillna('Unknown Region', inplace=True)


for tab in all_tabs:
    # Checking for required columns and concatenate
    if 'CODE' in tab.columns and 'NAME' in tab.columns and 'WHO_REGION' in tab.columns:
        temp_df = tab[['CODE', 'NAME', 'WHO_REGION']].dropna().drop_duplicates()
        countries_df = pd.concat([countries_df, temp_df], ignore_index=True)
    elif 'CODE' in tab.columns and 'NAME' in tab.columns:
        temp_df = tab[['CODE', 'NAME']].dropna().drop_duplicates()
        countries_df = pd.concat([countries_df, temp_df], ignore_index=True)

# This is the crucial step to ensure the 'CODE' column has no duplicates
countries_df.drop_duplicates(subset=['CODE'], keep='first', inplace=True)

final_tabs = [countries_df, tab1, tab2, tab3, tab4, tab5]


C:\Users\mrinm\AppData\Local\Temp\ipykernel_16276\874937203.py:32: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  countries_df['WHO_REGION'].fillna('Unknown Region', inplace=True)


In [80]:
def insert_into_tables(engine,tabs):
    """Inserts data from pandas DataFrames into SQL Server tables."""
    try:
        (countries_df, cleaned_coverage_df, cleaned_incidence_df,
         cleaned_reported_cases_df, cleaned_vaccine_intro_df,
         cleaned_vaccine_schedule_df) = tabs
        
        
        countries_df.to_sql('Countries', engine, if_exists='append', index=False)
        print("Data inserted into 'Countries' table successfully.")

        # Insert into other tables
        cleaned_coverage_df.to_sql('CoverageData', engine, if_exists='append', index=False)
        print("Data inserted into 'CoverageData' table successfully.")
        
        cleaned_incidence_df.to_sql('IncidenceRateData', engine, if_exists='append', index=False)
        print("Data inserted into 'IncidenceRateData' table successfully.")

        cleaned_reported_cases_df.to_sql('ReportedCasesData', engine, if_exists='append', index=False)
        print("Data inserted into 'ReportedCasesData' table successfully.")
        
        cleaned_vaccine_intro_df.to_sql('VaccineIntroduction', engine, if_exists='append', index=False)
        print("Data inserted into 'VaccineIntroduction' table successfully.")

        cleaned_vaccine_schedule_df.to_sql('VaccineScheduleData', engine, if_exists='append', index=False)
        print("Data inserted into 'VaccineScheduleData' table successfully.")


    except Exception as e:
        print(f"An error occurred during data insertion: {e}")

    finally:
        if engine: 
            engine.dispose()  # Close all open Connections


if __name__ == '__main__':
    try:
        engine = create_sql_engine()
        if engine:
            insert_into_tables(engine,final_tabs)
        else:
            print("Could not establish a database connection. Data insertion failed.")
    except FileNotFoundError:
        print("Error: One or more data files were not found. Please check file paths.")

SQLAlchemy engine created successfully.
Data inserted into 'Countries' table successfully.
Data inserted into 'CoverageData' table successfully.
Data inserted into 'IncidenceRateData' table successfully.
Data inserted into 'ReportedCasesData' table successfully.
Data inserted into 'VaccineIntroduction' table successfully.
Data inserted into 'VaccineScheduleData' table successfully.
